### ISSUE TITLE
for-of detection falsely triggered on typed arrays

### ISSUE THREAD

**Issue by mrkite** [palantir/tslint#3148]
### Bug Report

- __TSLint version__: 5.6.0
- __TypeScript version__: 2.4.2
- __Running TSLint via__:  CLI

#### TypeScript code being linted

```ts
let func = (data: Uint8Array): string => {
  let r: string = "";
  for (let i: number = 0; i < data.length; i++) {
    r += String.fromCharCode(data[i]);
  }
  return r;
};
```

with `tslint.json` configuration:

```json
{
  "defaultSeverity": "error",
  "extends": [
    "tslint:recommended"
  ],
  "jsRules": {},
  "rules": {},
  "rulesDirectory": []
}
```

#### Actual behavior

ERROR: test.ts[3, 3]: Expected a 'for-of' loop instead of a 'for' loop with this simple iteration

#### Expected behavior

No error.. you can't use for-of loops on Uint8Array or any Typed Arrays.

**Comment by IllusionMH** [comment]
Looks like problems with typings provided by TypeScript since TypedArrays do implement iterator protocol (according to MDN https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/@@iterator)

**Comment by ajafff** [comment]
Related to #2927 and basically a duplicate of #2021

> Probably flag may be added to suggest for..of only for arrays/strings.

That would require the rule to use the type checker. If it's a type checked rule, we could instead check if `--downlevelIteration` is enabled or `target >= ES6` in tsconfig.json and lint according to that setting. That wouldn't require a new option and you don't have to keep your configs in sync.

Unfortunately `TypeChecker` does not expose `isArrayLikeType` which would be needed to correctly implement this. Note that `no-for-in-array` currently has the same limitations: doesn't handle `ReadonlyArray` or shadowing the type `Array`
Exposing the necessary API(s) is tracked by https://github.com/Microsoft/TypeScript/issues/9879, https://github.com/Microsoft/TypeScript/issues/17680 and many other issues
